Replace InMemoryPersistedState with GatewayMetaState in CoordinatorTests

andrershov opened the pull request:
This PR replaces `InMemoryPersistedState` with`GatewayMetaState` in
`CoordinatorTests`.
When performing such a replacement the main question was: do we want to
emulate exceptions as we do today in `MockPersistedState` before
delegating to `GatewayMetaState` or do we want these exceptions to
propagate from the lower level, i.e. file system exceptions?
On the one hand, lower level exception propagation is already tested in
`GatewayMetaStateTests`, so this won't improve the coverage. 
On the other hand, the benefit of low-level exceptions is to see how all these
components work in conjunction.  Finally, we abandoned the idea of low-level
exceptions because we don't have a way to deal with `IOError` today in
`CoordinatorTests`, but hacking `GatewayMetaState` not to throw
`IOError` seems unnatural.
So `MockPersistedState` rarely throws an exception before delegating to
`GatewayMetaState`, which is not supposed to throw the exception.

This commit required two changes:
1. Move `GatewayMetaStateUT` to upper-level from
`GatewayMetaStatePersistedStateTests`, because otherwise it's not easy
to construct `GatewayMetaState` instance in `CoordinatorTests`.
2. Move addition of `STATE_NOT_RECOVERED_BLOCK` from `GatewayMetaState`
constructor to `GatewayMetaState.applyClusterUpdaters`, because
`CoordinatorTests` class assume that there is no such block and most of
them fail.

Other classes can be migrated from `InMemoryPersistedState` to
`GatewayMetaState` as well if requested by reviewers.

elasticmachine commented:
Pinging @elastic/es-distributed

DaveCTurner commented on server/src/test/java/org/elasticsearch/gateway/GatewayMetaStateUT.java:
```diff
@@ -0,0 +1,67 @@
+ * Metadata upgrade is tested in {@link  GatewayMetaStateTests} and different {@link ClusterStateUpdaters} in
+ * {@link ClusterStateUpdatersTests}.
+ */
+public class GatewayMetaStateUT extends GatewayMetaState {
```
Maybe `MockGatewayMetaState` would be a better name?

DaveCTurner commented on server/src/test/java/org/elasticsearch/gateway/GatewayMetaStateUT.java:
```diff
@@ -0,0 +1,67 @@
+
+    public GatewayMetaStateUT(Settings settings, NodeEnvironment nodeEnvironment,
+                              NamedXContentRegistry xContentRegistry,
+                       DiscoveryNode localNode) throws IOException {
```
nit: funny indentation.

ywelsch commented:
+1 on using it rarely

andrershov commented:
@DaveCTurner I've addressed your comments and changed `MockPersistedState` to use `GatewayMetaState` rarely and use `InMemoryPersistedState` in other cases. Ready for the 2nd pass.

andrershov commented:
run gradle build tests 1

andrershov commented:
run gradle build tests 2

andrershov commented:
@DaveCTurner Can you please make the next pass? I've fixed nits and last merge with master introduced node restart in CoordinatorTests (https://github.com/elastic/elasticsearch/pull/36897/commits/40b303cdd915098235704064ff844f59f25e2abc), so now `MockPersistedState` supports restarts when `GatewayMetaState` is used.

andrershov commented:
run gradle build tests

andrershov commented:
test this please

andrershov commented:
run gradle build tests 2
